### ISSUE TITLE
How is Resource Reaper signalled at the end of tests?

### ISSUE THREAD

**Issue by serefarikan** [testcontainers/testcontainers-dotnet#440]
Hi, 
I appreciate a question may not be an accepted issue for this project, so apologies in advance if that's the case here. 

I spent quite some time trying to understand how Resource Reaper is used to clean up containers and I fail to see how it kicks in at the end of tests. As far as I can see, Resource Reaper is run when a container is started, and it is given the containers to clean at that point. So what stops it from cleaning the containers before the tests are run? 

In other words, can you point me at the location in the code that ensures Resource Reaper is invoked only once the tests are complete? I can keep digging into the code but you'd be saving me a lot of time if you could just hint at the correct direction. 

Many thanks for such a nice library by the way!

**Comment by HofmeisterAn** [comment]
> So what stops it from cleaning the containers before the tests are run?

You don't know which resources you need to remove. Even if you label resources from previous runs, you don't know exactly what to remove. Especially if you run multiple tests parallel. In addition to that, it's better to remove resources right away and not within the next start. Removing old resources sounds easy, but there are a lot of use cases to consider.

> In other words, can you point me at the location in the code that ensures Resource Reaper is invoked only once the tests are complete?

You can run multiple Resource Reapers in different [sessions](https://github.com/HofmeisterAn/dotnet-testcontainers/blob/859826e98e619000ffe5029f088ed9db86946d6f/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L134), but there is always a default Resource Reaper through a [singleton](https://github.com/HofmeisterAn/dotnet-testcontainers/blob/859826e98e619000ffe5029f088ed9db86946d6f/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L84). Ryuk (the Docker image of the Resource Reaper) holds a network [connection](https://github.com/HofmeisterAn/dotnet-testcontainers/blob/859826e98e619000ffe5029f088ed9db86946d6f/src/DotNet.Testcontainers/Containers/ResourceReaper.cs#L294). If this connection is dropped (test process ends) Ryuk removes all resources within this session. Also take a look at the `MaintainRyukConnection` description.

**Comment by serefarikan** [comment]
@HofmeisterAn I went through the code after some coffee (magic ingredient ;) ) and just came back to ask if my understanding is correct and saw your answer. It's the infinite loop in `MaintainRyukConnection` isn't it? : Ryuk's behaviour is to keep a container alive unless it was queried/accessed for 10 seconds (hard coded in its source code btw, I think), so once the test is completed, Ryuk times out and removes the connections.  This is just the detail of course, your answer provides the context as well.

Many thanks for such a prompt response. Much appreciated.

**Comment by HofmeisterAn** [comment]
> Ryuk's behaviour is to keep a container alive unless it was queried/accessed for 10 seconds

To prevent misconceptions. .NET Testcontainers creates and holds a network connection to Ryuk. If this connection drops for more than 10 seconds, Ryuk starts to clean up.
